spark/src/test/scala-spark-master/org/apache/spark/sql/delta/DeltaVariantSuite.scala

```diff
@@ -126,4 +137,417 @@ class DeltaVariantSuite
+
+        sql(s"CONVERT TO DELTA parquet.`$path` $statsClause")
+        // Ensure Delta feature like column renaming works.
+        sql(s"ALTER TABLE delta.`$path` SET TBLPROPERTIES ('delta.columnMapping.mode' = 'name')")
+        sql(s"ALTER TABLE delta.`$path` RENAME COLUMN v TO new_v")
+
+        val expectedDf = spark.range(0, 100).selectExpr("parse_json(cast(id as string)) as new_v")
+        val actualDf = spark.sql(s"select * from delta.`$path`")
+        checkAnswer(actualDf, expectedDf.collect())
+      }
+    }
+  }
+
+  test("column mapping works") {
+    withTable("tbl") {
+      sql("""CREATE TABLE tbl USING DELTA
+          TBLPROPERTIES ('delta.columnMapping.mode' = 'name')
+          AS SELECT parse_json(cast(id as string)) v, parse_json(cast(id as string)) v_two
+          FROM range(5)""")
+      val expectedAnswer = spark.sql("select v from tbl").collect()
+
+      sql("ALTER TABLE tbl RENAME COLUMN v TO new_v")
+      checkAnswer(spark.sql("select new_v from tbl"), expectedAnswer)
+
+      sql("ALTER TABLE tbl DROP COLUMN new_v")
+      // 'SELECT *' from the test table should return the same as `expectedAnswer` because `v` and
+      // `v_two` are initially identical and `v` is dropped, resulting in a single column.
+      checkAnswer(spark.sql("select * from tbl"), expectedAnswer)
+    }
+  }
+
+  test("Variant can have default value set") {
+    withTable("tbl") {
+      sql("""CREATE TABLE tbl USING DELTA
+          TBLPROPERTIES ('delta.feature.allowColumnDefaults' = 'enabled')
+          AS SELECT parse_json(cast(id as string)) v from range(5)""")
+
+      sql("INSERT INTO tbl VALUES (DEFAULT)")
+      val nullCount = spark.sql("SELECT * FROM tbl WHERE v is null").count()
+      // Default DEFAULT value is null.
+      assert(nullCount == 1)
+
+      sql("ALTER TABLE tbl ALTER COLUMN v SET DEFAULT (parse_json('{\"k\": \"v\"}'))")
+      sql("INSERT INTO tbl VALUES (DEFAULT)")
+      val valCount = spark.sql("SELECT * FROM tbl WHERE v:k::string = 'v'").count()
+      // DEFAULT value is set to '{k: v}'.
+      assert(valCount == 1)
+    }
+  }
+
+  test("Variant can be used as a source for generated columns") {
+    withTable("tbl") {
+      DeltaTable.create(spark)
+        .tableName("tbl")
+        .addColumn("v", "VARIANT")
+        .addColumn(
+          DeltaTable.columnBuilder(spark, "vInt")
+            .dataType("INT")
+            .generatedAlwaysAs("v::int")
+            .build()
+        )
+        .execute()
+      spark.range(0, 100)
+        .selectExpr("parse_json(cast(id as string)) as v")
+        .write
+        .mode("append")
+        .format("delta")
+        .saveAsTable("tbl")
+      val expectedDf = spark.range(0, 100).selectExpr(
+        "parse_json(cast(id as string)) v",
+        "cast(id as int) vInt"
+      )
+      val actualDf = spark.sql("select * from tbl")
+      checkAnswer(actualDf, expectedDf.collect())
+    }
+  }
+
+  test("Variant cannot be created as a generated column") {
```

Comment:
It this something that we could eventually support?

Reply:
Yeah. Currently its not allowed because writers must be able to enforce that <variant value> <=> <generated expresion>. Currently, variants aren't comparable so its not allowed.

I updated the check in `GeneratedColumn` to be more specific. Before this PR, it was actually possible to `CAST(intVal as variant)` to generate a variant column. The new check completely disallows variant types.